catalog/secrets/api/src/main/java/org/projectnessie/catalog/secrets/Secret.java

```diff
@@ -15,5 +15,9 @@
  */
 package org.projectnessie.catalog.secrets;
 
+import java.util.Map;
+
 /** Base interface for all secrets. */
-public interface Secret {}
+public interface Secret {
```

Comment:
I am a bit worried by the current definition fo a secret as a collection of key value pairs. I think it would be simpler to think of it as a single key-value _entry_.

For example, `S3BucketOptions` now contain an `Optional<String> accessKey();` but nothing tells the user that the string must point to a secret containing not one, but two entries, one for the access key itself, another for the secret key. From a user perspective, this is imho difficult to get right and I expect users to struggle with this.

Also, thinking ahead of time of how this would map to the Kubernetes world, not all secrets are created with the shape expected by Nessie in this PR. E.g. if Nessie expects a `BasicCredentials` type, then the secret _must_ contain two entries named `name` and `secret`. This will become an operational burden since secrets may have been created by other applications (Helm charts, operators, ArgoCD etc.) with different shapes.

I would like to propose another way of referencing secrets: by URI. For example, `S3BucketOptions` could expose 2 methods: `Optional<URI> accessKey();` and `Optional<URI> secretAccessKey();`. The URI could have the following general form (to be refined):

```
nessie-secret://[secret-provider-id]/provider-specific-part
```

Examples:

```
nessie-secret://k8s/my-namespace/my-secret?key=secret-key
nessie-secret://hashicorp-vault/my-namespace/my-secret?key=secret-key
nessie-secret://file/mnt/config/passwd
nessie-secret://env/MY_PASSWORD
nessie-secret://quarkus-config/my.password
```

etc. Then, resolving a secret would boil down to fetching the provider by ID (must exist via ServiceLoader), then resolving the secret itself. 

Does that make sense?

Reply:
The purpose of having a composite is key rotation, which requires that composites like access-key+secret-access-key or username+password are in sync to prevent race conditions when reading (and caching!) secrets. AFAIK you can create secrets that have multiple key-value pairs in helm.

Reply:
Could actually use URNs in fact: `urn:nessie-secret:k8s:namespace:secret?key=secret-key` etc.

Reply:
> you can create secrets that have multiple key-value pairs in helm

Indeed but that's not what all charts do. Bitnami charts for databases e.g. usually create secrets only for the passwords, not for the users.

Reply:
Using `URI` to identify a secret is a good idea, so I've updated the branch.
Not sure about the query part or the URI though. But let's see how that would look once the follow-up #8707 is in. Maybe there's a solution for that - so users _could_ choose.

Reply:
Nice! I will have a look.

Re: secrets rotation. Could that be left to the secret provider as an implementation detail? IOW, if a provider is given a set of secrets to resolve, it could first group them by location, to avoid fetching the same resource more than once. And by doing so, the issue with out-of-sync secret pairs would be automatically solved.

Reply:
> > you can create secrets that have multiple key-value pairs in helm
> 
> Indeed but that's not what all charts do. Bitnami charts for databases e.g. usually create secrets only for the passwords, not for the users.

Well, the secrets-provider stuff here is (primarily) for object-store credentials. For database connections, we have to leverage what Quarkus, the connector implementation and the secrets-provider implementations have - it's not what we can control.

Reply:
> Nice! I will have a look.
> 
> Re: secrets rotation. Could that be left to the secret provider as an implementation detail? IOW, if a provider is given a set of secrets to resolve, it could first group them by location, to avoid fetching the same resource more than once. And by doing so, the issue with out-of-sync secret pairs would be automatically solved.

I haven't updated the branch. What I meant with "_could_ work" is a change on top of the direct follow-up - maybe sth that's inside the caching secrets provider - that _could potentially_ work.

Reply:
> Well, the secrets-provider stuff here is (primarily) for object-store credentials. For database connections, we have to leverage what Quarkus, the connector implementation and the secrets-provider implementations have - it's not what we can control.

Agreed. Was just noting. I don't know how object store secrets are usually created, but I think it's quite reasonable to assume that for AWS credentials, the access key and the secret key will be fetched from the same location. It's however less reasonable imho to assume the same for pairs of `(token, expires_at)` secrets. But let's see how it goes and if users complain or not.

Reply:
Updated to use the proposed URNs (via `URI`) and added a `ResolvingSecretsProvider`. It's probably never going to happen that someone uses different external secrets providers, but possible that some secrets are retrieved via Quarkus (pod environment) and some via an external secrets provider.